File: Data Structures & Algorithms/4sum/submission-0.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        n = len(nums)
        nums.sort()
        res = set()
        for i in range(n):
            for j in range(i+1, n):
                for k in range(j + 1, n):
                    for l in range(k + 1, n):
                        curr = nums[i] + nums[j] + nums[k] + nums[l]
                        if curr == target:
                            res.add((nums[i], nums[j], nums[k], nums[l]))



        return list(res)
        

File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {
    List<List<Integer>> res;
    List<Integer> current;
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        res = new ArrayList<>();
        current = new ArrayList<>();

        backtrack(0,current, nums, target);

        return (res);
    }
    private void backtrack(int i,List<Integer> current ,int[] nums, int target){
        if (i >= nums.length) {
            return;
        }
        int curr = 0;
        for (int num:current){
            curr += num;
        }
        if (curr > target){
            return;
        }
        if (curr == target){
            res.add(new ArrayList<>(current));
            return;
        }
        //Decision 1 - add itself
        current.add(nums[i]);
        backtrack(i, current, nums, target);
        // Decision 2 - add next number
        current.remove(current.size() -1);
        backtrack(i+1, current, nums, target);
    }
}
